File: bobo/lib/bobo/books/book.ex
defmodule Bobo.Books.Book do
  use Ecto.Schema
  import Ecto.Changeset

  schema "books" do
    field(:author, :string)
    field(:title, :string)
    field(:rating, :float)
    field(:genres, {:array, :string})
    field(:date_finished, :date)
    field(:comments, :string)

    timestamps()
  end

  @doc false
  def changeset(book, attrs) do
    book
    |> cast(attrs, [:title, :author, :rating, :genres, :date_finished, :comments])
    |> validate_required([:title, :author, :rating, :genres, :date_finished])
  end
end

File: bobo-ui/src/types/Book.ts
import { serialiseJSDate } from '../dateUtils/parse';

export default interface Book {
  id: number;
  title: string;
  author: string;
  genres: string[];
  rating: number;
  comments: string;
  updatedAt: string;
  dateFinished: string;
  __typename?: string;
}

export type NewBook = {
  title: string;
  author: string;
  genres: string[];
  rating: number;
  comments: string;
  dateFinished: string;
};

export const RequiredBookFields = [
  'title',
  'author',
  'genres',
  'rating',
  'dateFinished',
];

export const blankBook: Book = {
  id: 0,
  title: '',
  author: '',
  genres: [],
  rating: 3,
  comments: '',
  updatedAt: '',
  dateFinished: '',
};

export const withTodaysDate = (b: Book) => {
  return {
    ...b,
    dateFinished: serialiseJSDate(new Date()),
  };
};
